Adding ElastiCache Hook for creating, describing and deleting replication groups

VijayantSoni opened the pull request:
Adding ElastiCache Hook for creating, describing and deleting replication groups.
This hook interacts with AWS ElastiCache and supports following operations:

1. Creating replication groups
2. Deleting replication groups
3. Describing replication groups

Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

Commit:
Adding ElastiCache Hook for creating, describing and deleting replication group

Commit:
Fixing logging to not use format method, adding entry in docs for new hook

Commit:
Add doctring to all the classes where missing

Commit:
Fixing name of module in operators-and-hooks-ref.rst file

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,238 @@
+
+        super().__init__(client_type='elasticache', *args, **kwargs)
+
+    def __create_replication_group(self, config):
```
why are these private?

ashb replied:
different question: why are they even separate methods?

VijayantSoni replied:
My thought behind this is: this method (private) just does 1 thing which is making call to the 3rd party service and returns the response as is, no other decoration of any sort.

Then we have a wrapper on top of this which can do a lot of things on top of the response. In this case it is just checking on `status` of the replication group. 

But if we look at `delete_replication_group` API, it just a wrapper around `_delete_replication_group` which does all the work, retries and all, but in turn calls `__delete_replication_group` to make the 3rd party API call.

So keeping separate method I think helps to write more manageable and simple code, following Single Responsibility Principle.

But I do agree with your point that why are these private. My initial thought was, clients just use the public APIs and those handle everything, but maybe that thinking is not correct. Please provide your thoughts on this.

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,238 @@
+        :param config: Python dictionary to use as config for creating replication group
+        :return: Response from ElastiCache create replication group API
+        """
+        return self.get_conn().create_replication_group(**config)
```
use `self.conn` instead

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,238 @@
+            ElastiCacheStatus.STATUS_DELETING
+        ), status
+
+    def is_replication_group_available(self, replication_group_id):
```
should be called `wait_somerthing` since it polls/loops.

max time should be an argument to this fn too.

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,238 @@
+from airflow.providers.amazon.aws.hooks.base_aws import AwsBaseHook
+
+
+class ElastiCacheDescribeResponseKeys:
```
don't make these extra classes, just put the constants in ElastiCacheHook

ashb reviewed (changes requested):
This needs some unit tests adding too please.

(I have only given the code a cursory glance at this point)

VijayantSoni reviewed (commented):
Thanks @ashb , I will work on the changes.

Commit:
Adding unit tests, chaging to public APIs, adding poke params to wait methods

Commit:
Fix typo in varible name hook

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,259 @@
+        :param replication_group_id: ID of replication group to check for availability
+        :return: True if replication is available else False
+        """
+        sleep_time = initial_sleep_time or self.initial_poke_interval
```
This doesn't let you put any of these values as `0` -- it would then fallback to the `self.initial_poke_interval`

ashb replied:
Oh maybe that is fine given the behaviour of them.

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,259 @@
+            replication_group_id=replication_group_id
+        ) == self.STATUS_AVAILABLE
+
+    def should_stop_poking(self, replication_group_id):
```
```suggestion
    def _should_stop_poking(self, replication_group_id):
```

ashb replied:
`_has_reached_terminal_state` would be a more descriptive name.

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,259 @@
+        Call ElastiCache API for creating a replication group
+        :param config: Python dictionary to use as config for creating replication group
+        :return: Response from ElastiCache create replication group API
+        """
```
All of your doc comments need a blank line between prose and params, else it won't render correctly
```suggestion
        """
        Call ElastiCache API for creating a replication group
        
        :param config: Python dictionary to use as config for creating replication group
        :return: Response from ElastiCache create replication group API
        """
```

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,259 @@
+        """
+        return self.get_replication_group_status(
+            replication_group_id=replication_group_id
+        ) == self.STATUS_AVAILABLE
```
```suggestion
        return self.get_replication_group_status(replication_group_id) == self.STATUS_AVAILABLE
```

It's rather needlessly verbose to use a named argument for a function with only takes one parameter.

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,259 @@
+            self.log.info('Status : %s', status)
+
+            if stop_poking:
+                self.log.info('Received signal to stop poking. Current status : "%s"', status)
```
We don't need this log like, we just said what it is.

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,259 @@
+                    )
+
+                elif re.search(self.RESPONSE_DELETING_STATE, message):
+                    self.log.info(
```
Wait, we try and delete it every time around the loop?

That seems.... odd. Why not 1) issue delete request, 2) describe/sleep until it is eventually deleted.

VijayantSoni replied:
Because we can't delete from any state but only `available`, hence I issued `delete` each time. I have changed it to get status first and then only issue a `delete` when status is `available`.

ashb reviewed (changes requested):
See individual comments

Commit:
Changing behaviour of delete to make use of describe API after issuing a delete request, doc string formaating, remove redundant log, minor refactor

Commit:
Handling different status when making a delete call

Commit:
Fix docstring and line-too-long issue

Commit:
Change to multiple line comments from docstring

VijayantSoni committed:
Updating test cases for mocking elasticache responses

VijayantSoni committed:
Fix status name in comments in  method

ashb reviewed (commented):
I'm not sure an operator like this belongs in Aiflow.

Yes, it could, but there are better tools for managing (creating/deleting) resources such as HashiCorp's Terraform, or AWS CloudFormation (which already has an operator in master).

Can you describe in more detail why you want this please?

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+        :param replication_group_id: ID of replication group to check for status
+        :return: Current status of replication group
+        """
+        return self.describe_replication_group(replication_group_id)[self.REPLICATION_GROUPS][0][self.STATUS]
```
```suggestion
        return self.describe_replication_group(replication_group_id)["ReplicationGroups"][0]["Status"]
```

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+    STATUS_DELETING = 'deleting'
+    STATUS_CREATE_FAILED = 'create-failed'
+    STATUS_MODIFYING = 'modifying'
+    STATUS_SNAPSHOTTING = 'snapshotting'
```
```suggestion
```

These constants make it harder to understand the code and isn't the style we use in Airfow, please use the string literals directly.

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+        :param replication_group_id: ID of replication group to check for availability
+        :return: True if available else False
+        """
+        return self.get_replication_group_status(replication_group_id) == self.STATUS_AVAILABLE
```
```suggestion
        return self.get_replication_group_status(replication_group_id) == "available"
```

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+            self.STATUS_AVAILABLE,
+            self.STATUS_CREATE_FAILED,
+            self.STATUS_DELETING
+        ), status
```
```suggestion
        return status in (
            "available",
            "create-failed",
            "deleting"
        ), status
```

This one could be worth extracting to a class variable:

```python
    TERMINAL_STATES = frozenset({"available", "create-failed", "deleting"})
```

And then this becomes

```suggestion
        return status in self.TERMINAL_STATES, status
```

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+        max_retries=None
+    ):
+        """
+        Check if replication is available or not by performing a describe over it
```
This doc talks about replication -- but is it just health of elasticache? (Because this still applies when a single node EC is used, right?)

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+        exponential_back_off_factor = exponential_back_off_factor or self.exponential_back_off_factor
+        max_retries = max_retries or self.max_retries
+        num_retries = 0
+        status = self.STATUS_CREATE_FAILED
```
```suggestion
        status = "not-found"
```

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+
+            sleep_time = sleep_time * exponential_back_off_factor
+
+        if status != self.STATUS_AVAILABLE:
```
```suggestion
        if status != "available":
```

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+        :param max_retries: Max tries for checking availability of replication group
+        :param exponential_back_off_factor: Factor for deciding next sleep time
+        :param initial_sleep_time: Initial sleep time in seconds
+        :param replication_group_id: ID of replication group to check for availability
```
Could you add types to these please?

```suggestion
        :param max_retries: Max tries for checking availability of replication group
        :type max_retries: int
        :param exponential_back_off_factor: Factor for deciding next sleep time
        :type exponential_back_off_factor: ...
        :param initial_sleep_time: Initial sleep time in seconds
        :type initial_sleep_time:
        :param replication_group_id: ID of replication group to check for availability
        :type replication_group_id:
```

Same for all the other public methods too please.

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+                    response = self.delete_replication_group(replication_group_id=replication_group_id)
+
+            except self.conn.exceptions.ReplicationGroupNotFoundFault:
+                self.log.info("Replication group with ID : '%s' does not exist", replication_group_id)
```
```suggestion
                self.log.info("Replication group with ID '%s' does not exist", replication_group_id)
```

ashb commented on airflow/providers/amazon/aws/hooks/elasticache.py:
```diff
@@ -0,0 +1,265 @@
+
+                message = exp.response['Error']['Message']
+
+                self.log.info('Received message : %s', message)
```
```suggestion
                self.log.error('Received error message from AWS API: %s', message)
```

ashb replied:
(Or maybe at warning?)

ashb commented on tests/providers/amazon/aws/hooks/test_elasticache.py:
```diff
@@ -0,0 +1,188 @@
+        response = self.hook.create_replication_group(config=self.replication_group_config)
+
+        assert response["ReplicationGroup"]["ReplicationGroupId"] == self.replication_group_id
+        assert response["ReplicationGroup"]["Status"] == "creating"
```
This isn't testing real code here -- you are only calling a mock function but not actually calling the real code anywhere from the hook.

ashb reviewed (changes requested):
Your tests need some work still -- you aren't actually importing the hook you want to test in your new test file, so you haven't tested any of the code.

You probably want to do so something like:


```
    def setUp(self):
        self.hook = ElastiCacheHook()
        setattr(self.hook, 'conn', Mock())`
```

and then set various properties/return values of self.hook.conn (similar to how you are doing on `self.hook`) the goal being to test all the code in your Hook class, but not reach out to an actual AWS API.

I won't be able to re-review this again for 6 weeks, so someone you'll have to ask in slack for a re-review -- I'd suggest one of @kaxil, @mik-laj or @potiuk.

VijayantSoni committed:
Renaming hook as ElastiCacheReplicationGroupHook because there are other types as well within ElasatiCache
Removed separate string constants for API response and replaced with hardcoded strings everywhere
Fixed tests to mock hook.conn and not hook, updated tests accordingly
changed semantics of max_retries for polling, now it means number of retries excluding first try
max_retries = 2 will mean 2 retires after the initial try

VijayantSoni committed:
Change 'not >=' to '<' as per pylint error.
Fix name of test file to correspond with hook file name

feluelle commented on docs/operators-and-hooks-ref.rst:
```diff
@@ -446,6 +446,12 @@ These integrations allow you to perform various operations within the Amazon Web
 
+   * - `Amazon ElastiCache <https://aws.amazon.com/elasticache/redis//>`__
+     -
+     -
+     - :mod:`airflow.providers.amazon.aws.hooks.elasticache_replication_group`
```
```suggestion
     - :mod:`airflow.providers.amazon.aws.hooks.elasticache_replication_group`
     -
```

VijayantSoni replied:
Fixed

feluelle commented on tests/providers/amazon/aws/hooks/test_elasticache_replication_group.py:
```diff
@@ -0,0 +1,357 @@
+class TestElastiCacheHook(TestCase):
+    """
+    Test ElastiCacheHook
+    """
```
```suggestion
```
You don't need doc strings for test classes I think. However if you still want to document it, please add more meaningful message.

VijayantSoni replied:
Removed.

feluelle commented on tests/providers/amazon/aws/hooks/test_elasticache_replication_group.py:
```diff
@@ -0,0 +1,357 @@
+    def test_get_conn_not_none(self):
+        """
+        Test connection is not None
+        """
```
```suggestion
```
Same for the methods.

VijayantSoni replied:
Removed.

feluelle commented on tests/providers/amazon/aws/hooks/test_elasticache_replication_group.py:
```diff
@@ -0,0 +1,357 @@
+
+    def setUp(self):
+        self.hook = ElastiCacheReplicationGroupHook()
+        setattr(self.hook, 'conn', Mock())
```
```suggestion
        self.hook.conn = Mock()
```

VijayantSoni replied:
It gives following IDE warning after this change:
```
Property 'conn' cannot be set
Inspection info: This inspection checks that properties are accessed correctly: read-only not set, write-only not read, non-deletable not deleted
```
will add suppression for the statement as well.

feluelle commented on tests/providers/amazon/aws/hooks/test_elasticache_replication_group.py:
```diff
@@ -0,0 +1,357 @@
+        """
+        Test connection is not None
+        """
+        self.assertIsNotNone(self.hook.conn)
```
```suggestion
        assert self.hook.conn is not None
```

VijayantSoni replied:
done

feluelle commented on airflow/providers/amazon/aws/hooks/elasticache_replication_group.py:
```diff
@@ -0,0 +1,289 @@
+
+        while not stop_poking and num_tries <= max_retries:
+            stop_poking, status = self._has_reached_terminal_state(replication_group_id=replication_group_id)
+
```
```suggestion
            status = self.get_replication_group_status(replication_group_id=replication_group_id)
            stop_poking = status in self.TERMINAL_STATES
```

feluelle commented on airflow/providers/amazon/aws/hooks/elasticache_replication_group.py:
```diff
@@ -0,0 +1,289 @@
+        """
+        status = self.get_replication_group_status(replication_group_id=replication_group_id)
+
+        return status in self.TERMINAL_STATES, status
```
```suggestion
```
WDYT of removing this and being explicit about it (see line 146 & 147)

feluelle replied:
Or pass status to this function and return only if it is a terminal state. Returning two values where one depends on the other one looks to me that this should be two statements like I suggested in line 146 & 147.

VijayantSoni replied:
I think it makes sense. Since it was only getting used at one place, we can write 2 statements (as you suggested) and remove that method which was returning multiple values.
Now that you pointed out, I think it was not correct because the method name was `_has_reached_terminal_state` but it was returning `status` also along with the boolean flag which I think was not correct.

feluelle commented on airflow/providers/amazon/aws/hooks/elasticache_replication_group.py:
```diff
@@ -0,0 +1,289 @@
+
+                sleep(sleep_time)
+
+                sleep_time = sleep_time * exponential_back_off_factor
```
```suggestion
                sleep_time *= exponential_back_off_factor
```

feluelle commented on airflow/providers/amazon/aws/hooks/elasticache_replication_group.py:
```diff
@@ -0,0 +1,289 @@
+
+                sleep(sleep_time)
+
+                sleep_time = sleep_time * exponential_back_off_factor
```
```suggestion
                sleep_time *= exponential_back_off_factor
```

feluelle reviewed (commented):
Welcome @VijayantSoni! :)

VijayantSoni committed:
Remove _has_reached_terminal_state method
Some python cosmetic changes
Replace occurences of elasticache-hook with elasticache-replication-group-hook
Remove doc strings from tests
Fix formatting in operators-and-hooks-ref.rst

feluelle commented on airflow/providers/amazon/aws/hooks/elasticache_replication_group.py:
```diff
@@ -0,0 +1,277 @@
+        :type exponential_back_off_factor: float
+        :param initial_poke_interval: Initial sleep time in seconds
+        :type initial_poke_interval: float
+        """
```
Please move this to the `ElastiCacheReplicationGroupHook` class doc string.

feluelle commented on tests/providers/amazon/aws/hooks/test_elasticache_replication_group.py:
```diff
@@ -0,0 +1,291 @@
+
+        self.hook.conn.exceptions.ReplicationGroupNotFoundFault = MockReplicationGroupNotFoundFault
+
+        raise self.hook.conn.exceptions.ReplicationGroupNotFoundFault
```
You don't need to save `describe_call_count_for_delete`. `side_effect` will do that for you.

Example from https://docs.python.org/3/library/unittest.mock.html#quick-guide:
```python
>>> mock.side_effect = [5, 4, 3, 2, 1]
>>> mock(), mock(), mock()
(5, 4, 3)
```

feluelle replied:
So you could do sth. like this: 
```python
mock.side_effect = [{
    "ReplicationGroups": [{
        "ReplicationGroupId": self.REPLICATION_GROUP_ID,
        "Status": "available"
    }]
}, {
    "ReplicationGroups": [{
        "ReplicationGroupId": self.REPLICATION_GROUP_ID,
        "Status": "deleting"
    }]
}, self.hook.conn.exceptions.ReplicationGroupNotFoundFault()]
```

VijayantSoni replied:
Thanks @feluelle, this makes it better.
The only issue I faced is that the last item has to be an exception of type BaseException so I have created a separate method for that and returning that method from here.

VijayantSoni committed:
Removing custom tracking of calls to mock and leveraging mock's tracking feature
Moved doc string from __init__ method to class doc string

feluelle commented on airflow/providers/amazon/aws/hooks/elasticache_replication_group.py:
```diff
@@ -0,0 +1,279 @@
+    ):
+        """
+
+        """
```
```suggestion
```
You don't need that. The linter should not complain if you don't have that.

VijayantSoni replied:
my bad, missed this.

feluelle commented on airflow/providers/amazon/aws/hooks/elasticache_replication_group.py:
```diff
@@ -0,0 +1,279 @@
+
+        if not deleted:
+            raise AirflowException(
+                'Replication group could not be deleted. Response "{0}"'.format(response)
```
```suggestion
                f'Replication group could not be deleted. Response "{response}"'
```

VijayantSoni committed:
Remove redundant doc string quotes, change to f-strings for string formating (only 1 place)

ashb reviewed (commented):
~There's already an aws.hooks.elasticache isn't there? I'd have expected this to just be new methods on the that class -- RGs are a "subset"/feature of Elasticache basically.~

Ignore this. I thought this PR had already landed and was confusing my memory of this PR with... this PR.

ashb commented on airflow/providers/amazon/aws/hooks/elasticache_replication_group.py:
```diff
@@ -0,0 +1,276 @@
+        :param initial_sleep_time: Initial sleep time in seconds
+        :type initial_sleep_time: float
+        :param replication_group_id: ID of replication group to check for availability
+        :type replication_group_id: str
```
Can you have doc order match function signature order.

And mention that back_off and sleep_time default to class level values if not passed.

VijayantSoni committed:
Fixing order of params in function doc string to match function signature
Adding default behaviour of params to doc string

VijayantSoni committed:
Running black on hook file and test file to resolve static check failures

VijayantSoni committed:
resolving static checks for test cases

feluelle reviewed (approved):
LGTM @VijayantSoni 👍 

Sorry for the late reply.
